Creating keras develop branch

callahantiff opened the pull request:
- Pushing updates to `word2vec,py`, `cbow_list_batcher.py` and `test_word2vec.py` scripts with refactoring of syntax (i.e. Google Style guides) and cleaning of a few bugs
- Passes all tests and `pylint`
- Not including refactoring of specific methods yet -- will push these changes in next iteration, so please don't yet delete branch

justaddcoffee commented:
Google style guide we are going to follow:
http://google.github.io/styleguide/pyguide.html

callahantiff commented:
looking into why it's failing the `test_skip_gram_word2vec.py` tests now.

justaddcoffee commented:
@callahantiff made a merge commit, hopefully didn't break anything for you, please check

callahantiff commented:
> @callahantiff made a merge commit, hopefully didn't break anything for you, please check

Seems to be OK! 👍

callahantiff commented:
Fixed failing tests, but now it looks to be failing on coverage. No tests were deleted in the changes I made.

justaddcoffee commented:
> Fixed failing tests, but now it looks to be failing on coverage. No tests were deleted in the changes I made.

Coveralls seems to be complaining because more coded was added without more tests? (really strict!)
https://coveralls.io/builds/28722595/source?filename=xn2v/word2vec.py

callahantiff commented:
> > Fixed failing tests, but now it looks to be failing on coverage. No tests were deleted in the changes I made.
> 
> Coveralls seems to be complaining because more coded was added without more tests? (really strict!)
> https://coveralls.io/builds/28722595/source?filename=xn2v/word2vec.py

How interesting, I didn't actually add anything new just refactored some of the existing methods. I wonder if the increase in rows from my changes file triggered this?

Regardless, thanks for sending this. I'm happy to try and address this now or I can hold off until we get to this part of our planned refactoring, since I assume there will be a bunch of new changes.

justaddcoffee commented:
@callahantiff want to have a crack at merging the latest `develop` with some changes from Peter/Vida? Looks like `creating_keras_develop_branch` touches a lot of the same places in  `xn2v/w2v/cbow_list_batcher.py` and `xn2v/word2vec.py`

callahantiff commented:
> @callahantiff want to have a crack at merging the latest `develop` with some changes from Peter/Vida? Looks like `creating_keras_develop_branch` touches a lot of the same places in `xn2v/w2v/cbow_list_batcher.py` and `xn2v/word2vec.py`

@justaddcoffee - Happy to do that. Just to confirm we are on the same page, I should focus on trying to fix the merge conflicts that resulted from the changes made to the `develop` branch today?

justaddcoffee commented:
> @justaddcoffee - Happy to do that. Just to confirm we are on the same page, I should focus on trying to fix the merge conflicts that resulted from the changes made to the `develop` branch today?

Yep, that's what I was thinking - just merge what Peter and Vida put in `develop` today, since it might affect how we proceed on this branch...

callahantiff commented:
> > @justaddcoffee - Happy to do that. Just to confirm we are on the same page, I should focus on trying to fix the merge conflicts that resulted from the changes made to the `develop` branch today?
> 
> Yep, that's what I was thinking - just merge what Peter and Vida put in `develop` today, since it might affect how we proceed on this branch...

Okie doke, no problem. I'm out of lab in 30 min and will get this done first thing. Sorry you didn't hear much from me today. I plan to make up for it tomorrow!

justaddcoffee commented:
> Okie doke, no problem. I'm out of lab in 30 min and will get this done first thing. Sorry you didn't hear much from me today. I plan to make up for it tomorrow!

No problem!

callahantiff commented:
OK, @justaddcoffee , we should be good to go.

Here is what I did:
- Fixed all merge conflicts  
- Added missing import statement for `logging.handlers`, function was not be reachable by `import logging`. Adding `import logging.handlers` fixed the problem. Note, using `from logging import handlers` should have also resolved this, but was triggering an additional error. 
- Resolved several errors caused by tests trying to import classes that had been remove (i.e. the `WeightedTriple`, `StringInteraction`, and `xn2vParser` classes (see removal commit [here](https://github.com/monarch-initiative/N2V/commit/31b353ce61a42f67249a273b277635da9d3730610))).  The tests, which only tested the functionality of these classes have been removed.

justaddcoffee commented:
@callahantiff thanks!

callahantiff commented:
Just wanted to share in case anyone else gets this error from Codacy. Codacy was flagging the following line from `text_encoder.py` as potentially containing a password:

```python
tokenizer = Tokenizer(num_words=max_vocab_size + 1, oov_token='UNK', lower=True,
                              filters='\'!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n')
```

In particular, it was not happy about the `oov_token='UNK` parameter, thinking that we were passing it a hardcoded password (i.e. `"UNK"`). I was able to silence the error by changing the value passed to this parameter to `oov_token=['UNK][0]`. I wouldn't say this is the best solution, but it works for now. If anyone knows of a better solution (i.e. updating Codacy), please let me know and I am happy to edit accordingly!

pnrobinson commented:
The workaround seems fine. It is also possible to turn off the rule that leads to the warning on the Codacy website directly.

callahantiff commented:
> The workaround seems fine. It is also possible to turn off the rule that leads to the warning on the Codacy website directly.

Thanks Peter! I saw that too, but was hesitant to turn it off. In general, I think that kind of a check is a really great idea. If you are OK with my workaround, I say we leave it for now. I'll make a comment in the code to remind myself of our decision 😸.

callahantiff commented:
@pnrobinson and @vidarmehr - I'm working on the tensorization portion of the refactor and wanted to verify some base assumptions wrt to the expected inputs for our graph embedding models (i.e. node2vec model implementations like skipgrram and cbow that are run on a graph).

Is it a fair assumption that we should be able to handle walks (or other types of output derived from a graph) of differing lengths? Our current code outputs walks of the same length, but I assume we would want to be able to handle a variety of input? The cool part about this refactor is that there are robust builtin `tf` tools that will work for either scenario, but I want to be smart about the design so would like to hear your thoughts before I make these changes.

I will commit the updated `text_encoder.py` script tonight refactoring it prompted my question.

pnrobinson commented:
Right now the walks all have the same length, but there is no need to enforce that and maybe we will want variable walks for some future algorithms!


---

The information in this email, including attachments, may be confidential and is intended solely for the addressee(s). If you believe you received this email by mistake, please notify the sender by return email as soon as possible.

callahantiff commented:
> Right now the walks all have the same length, but there is no need to enforce that and maybe we will want variable walks for some future algorithms!
> […](#)
> --- The information in this email, including attachments, may be confidential and is intended solely for the addressee(s). If you believe you received this email by mistake, please notify the sender by return email as soon as possible.

Excellent, I will proceed accordingly. I'll push the easier changes first (i.e. changes to output from `csf_graph.py` and `hetnode2vec.py` and then work through the the w2v models.

callahantiff commented:
@justaddcoffee - Thanks for meeting with me today. From our discussion, we decided that we would split up the next step of the refactoring process such that I will work on the `word2vec_tf.py` script (and associated tests, adding `_tf` to the filename) and you will work on the `hetnode2vec_tf.py` script (and associated tests, adding `_tf` to the filename). Note that I committed a `hetnode2vec_tf.py` script for you so you should see that fiel in this branch.

That work is now done, but before I shown you the changes (below), I wanted to remind you how the tensors are created:  
- Tensor output will come in two types depending on the structure of its input:  
  - if the input data is a single list --> `tf.data.Dataset`  
  - if the input data is a list of lists where all inner lists have the same length --> `tf.data.Dataset`  
  - if the input data is a list of lists, but the inner lists have differing lengths --> `tf.RaggedTensor`  
- Tensor Object Creation:  
  - `tf.data.Dataset` objects are created by passing the input data to `tf.data.Dataset.from_tensor_slices()`  
  - `tf.RaggedTensor` objects are created by passing the input data to `tf.ragged.constant()`  

<br>

With that in mind, I updateed the output of two methods from the `csf_graph_tf.py` script. Specifically, I needed to update convert the following functions:
1. `edges()`:  
    _ORIGINAL OUTPUT_ --> `List[List]`  
    ```python
    [('d1', 'd2'),
    ('d1', 'd3'),
    ('d1', 'g3'),
    ('d2', 'd1'),
    ('d2', 'd3')]
    ```
    _`tf` OUTPUT_  --> `tf.data.Dataset` 
    ```python
    tf.Tensor([b'd1' b'd2'], shape=(2,), dtype=string)
    tf.Tensor([b'd1' b'd3'], shape=(2,), dtype=string)
    tf.Tensor([b'd1' b'g3'], shape=(2,), dtype=string)
    tf.Tensor([b'd2' b'd1'], shape=(2,), dtype=string)
    tf.Tensor([b'd2' b'd3'], shape=(2,), dtype=string)
    ```
2. `edges_as_ints()`:  
    _ORIGINAL OUTPUT_ -->  `List[List]`  
    ```python
    [(0, 1),
     (0, 2),
     (0, 5),
     (1, 0)]
    ```
    _`tf` OUTPUT_    --> `tf.data.Dataset`  
    ```python
    tf.Tensor([0 1], shape=(2,), dtype=int32)
    tf.Tensor([0 2], shape=(2,), dtype=int32)
    tf.Tensor([0 5], shape=(2,), dtype=int32)
    tf.Tensor([1 0], shape=(2,), dtype=int32)
    ```
<br>

**Other Things to Note:**   
- All refactored scripts are duplicated and renamed by appending `_tf` to the filename. To work easily with fewest changes during refactoring, I have also been changing the import statements in the test files so that we don't have to edit the `__init__.py` scripts until the full refactor is done (e.g. changing `from xn2v import CSFGraph` to `from xn2v.csf_graph.csf_graph_tf import CSFGraph`)  
- To get the length of a tensor called `data`:  
  - if tensor is a `tf.RaggedTensor` --> `data.shape[0]`
  - if tensor is a `tf.data.Dataset` --> `len([edge for edge in data])` 

<br>

**Helpful `tf` Resources:**  
- [Medium Post](https://medium.com/@prasad.pai/how-to-use-dataset-and-iterators-in-tensorflow-with-code-samples-3bb98b6b74ab)  
- [Adventures in Machine Learning Post](https://adventuresinmachinelearning.com/tensorflow-dataset-tutorial/)  

<br>

Hope that is helpful, let me know if you have any questions! 😸 

Also, I will likely word on the `word2vec.py` script changes on Sunday. If anything changes, I will let you know.

justaddcoffee commented:
@callahantiff thanks! I'll start on this now-ish

justaddcoffee commented on xn2v/csf_graph/csf_graph_tf.py:
```diff
@@ -0,0 +1,395 @@
+        else:
+            return False
+
+    def edges(self) -> Union[tf.data.Dataset, tf.RaggedTensor]:
```
@callahantiff do we need to allow `tf.RaggedTensor`s? IIUC, all edges will be of length 2 in `edges()` and `edges_as_ints()` right?

callahantiff replied:
While the code we current have does not need to account for `tf.RaggedTensors`, I talked to Peter about it and for future iterations and/or as new path finding or embedding methods are added we don't want to be constrained to lists or paths of the same lengths.

This is why I am proceeding with the refactor with the assumption that we should be able to accommodate either list of paths of equal length (`tf.data.Dataset`) of lists of paths of unequal lengths (`tf.RaggedTensor`). Does that make sense?

Are you running into a problem?

justaddcoffee replied:
Ah okay, thanks! That makes sense, no problem supporting `RaggedTensors`, just trying to understand the code better

callahantiff replied:
Sounds good. Is our plan still the same as Friday? I will focus on the `word2vec_tf.py` script and tests and you are working on the `xn2v/hetnode2vec_tf.py` script and tests?

I also noticed there are a bunch of merge conflicts, I'm happy work on those later today.

justaddcoffee replied:
Yep, I'm working on `xn2v/hetnode2vec_tf.py` - hope to have some more code soon

callahantiff replied:
Sounds Great! 😎

callahantiff commented:
@justaddcoffee - I just resolved all of the merge conflicts between this branch and the `develop` branch. Some of the code did change as a result of this. You may want to re-copy the content from `hetnode2vec.py` to `hetnode2vec_tf.py`.

callahantiff commented:
> Thanks so much -- I will merge now but we will need to continue polishing!

Sounds great @pnrobinson - Happy to help get the tests updated. Perhaps we can discuss a plan for this during this week's meeting?
